### PR TITLE
idle: enable the compositors to add custom idle timeouts

### PR DESCRIPTION
**Pull request by ammen99**
Is something like this desirable in wlroots? I want it so that I can set DPMS from inside Wayfire (Wayfire doesn't have swaymsg and so a plugin must turn off the outputs).

### CODE REVIEW

File: types/wlr_idle.c [hunk 1]
```diff
@@ -246,3 +216,65 @@ struct wlr_idle *wlr_idle_create(struct wl_display *display) {
 void wlr_idle_notify_activity(struct wlr_idle *idle, struct wlr_seat *seat) {
 	wlr_signal_emit_safe(&idle->events.activity_notify, seat);
 }
+
+struct wlr_idle_timeout *wlr_idle_timeout_create(struct wlr_idle *idle,
+	struct wlr_seat *seat, uint32_t timeout) {
```
Comment (ddevault) [inline types/wlr_idle.c]: style: indent the second line by two indents

File: include/wlr/types/wlr_idle.h [hunk 2]
```diff
@@ -68,4 +73,15 @@ void wlr_idle_notify_activity(struct wlr_idle *idle, struct wlr_seat *seat);
  */
 void wlr_idle_set_enabled(struct wlr_idle *idle, struct wlr_seat *seat,
 	bool enabled);
+
+/**
+ * Create a new timer on the given seat. The fired event will be called after
```
Comment (ddevault) [inline include/wlr/types/wlr_idle.h]: Do you mean the idle event?

**Comment by ddevault** [comment]
+1

File: include/wlr/types/wlr_idle.h (hunk 2 above)
Comment (ammen99) [reply include/wlr/types/wlr_idle.h]: Oh yeah, I initially called this `fired` then decided `idle` is a better name

**Comment by ammen99** [comment]
Updated to address the comments from @ddevault

File: types/wlr_idle.c [hunk 3]
```diff
@@ -246,3 +216,65 @@ struct wlr_idle *wlr_idle_create(struct wl_display *display) {
...
+		free(timer);
+
+		return NULL;
+	}
+
+	if (timer->enabled) {
+		// arm the timer
+		wl_event_source_timer_update(timer->idle_source, timer->timeout);
+		if (timer->timeout == 0) {
+			idle_notify(timer);
```
Comment (emersion) [inline types/wlr_idle.c]: This won't work, `timer->resource` is NULL at this point.

I guess the solution will be to keep a private constructor with a `resource` argument, and expose a public constructor which provides `resource = NULL` to the private constructor.

File: include/wlr/types/wlr_idle.h [hunk 4]
```diff
@@ -50,6 +50,11 @@ struct wlr_idle_timeout {
 	struct wl_listener seat_destroy;
 
 	void *data;
+	struct {
+		struct wl_signal idle;
+		struct wl_signal resume;
+		struct wl_signal destroy;
+	} events;
```
[the remark this answers is not in the document]
Comment (emersion) [reply include/wlr/types/wlr_idle.h]: Hmm. Seemed pretty consistent to me. Do you have examples in mind?

The idea is to move the "internal" things at the end of the struct.

[the remark this answers is not in the document]
Comment (ammen99) [reply include/wlr/types/wlr_idle.h]: Ah ok.
What I have seen is for example foreign-toplevel (granted, it was implemented by me :)
But anyway, data comes usually after the events.

**Comment by ammen99** [comment]
@emersion Updated with a private constructor, hope I haven't missed anything this time.

File: types/wlr_idle.c [hunk 5]
```diff
@@ -101,34 +100,44 @@ static void handle_input_notification(struct wl_listener *listener, void *data)
...
-		return;
+		if (idle_resource) {
+			wl_resource_post_no_memory(idle_resource);
+		}
+		return NULL;
+	}
+
+	if (idle_resource)
+	{
+		timer->resource = wl_resource_create(
```
Comment (emersion) [inline types/wlr_idle.c]: Hmm. Can we just create the resource in the caller, and `wl_resource_set_user_data(resource, timer)` here?

Comment (ammen99) [reply types/wlr_idle.c]: You're right, makes more sense. I totally forgot I can just check user data later.

**Comment by emersion** [comment]
Thanks!

**Comment by ammen99** [comment]
Thanks for the reviews!

**Pull request merged**